From 86c8074609f9d9c303536fa950339c7b5cbad489 Mon Sep 17 00:00:00 2001
From: Ken Ma <make@marvell.com>
Date: Sat, 11 Mar 2017 00:53:38 +0800
Subject: [PATCH 0901/1345] xhci-plat: add phy power opeartion in PM

commit  c89367956524a11595a320b90673294ce4658df3 from
https://github.com/MarvellEmbeddedProcessors/linux-marvell.git

- Add PHY power off/on to suspend/resume;
- XHCI's main hcd is for USB2 while its shared hcd is for USB3; If phys
  are separated for USB2 and USB3, then main hcd's phy represents the
  dedicated USB2 phy while shared hcd's phy represents the USB3 phy
  which is a different phy than main hcd, both phys must be powered
  off/on; otherwise the two hcds shares a same phy which is for both
  USB2 and USB3, this phy only needs to be power off/on once.

Change-Id: I9f3084496acc00d05fc6ab5c90393bbcddfc228e
Signed-off-by: Terry Zhou <bjzhou@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/34903
Reviewed-by: Victor Gu <xigu@marvell.com>
Tested-by: Victor Gu <xigu@marvell.com>
Signed-off-by: Ken Ma <make@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/37372
Tested-by: iSoC Platform CI <ykjenk@marvell.com>
Reviewed-by: Omri Itach <omrii@marvell.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/usb/host/xhci-plat.c |   53 ++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 53 insertions(+)

diff --git a/drivers/usb/host/xhci-plat.c b/drivers/usb/host/xhci-plat.c
index 5025359..695787e 100644
--- a/drivers/usb/host/xhci-plat.c
+++ b/drivers/usb/host/xhci-plat.c
@@ -432,6 +432,24 @@ static int __maybe_unused xhci_plat_suspend(struct device *dev)
 	 */
 	ret = xhci_suspend(xhci, device_may_wakeup(dev));
 
+	/*
+	* xhci's main hcd is for USB2 while its shared hcd is for USB3;
+	* If phys are separated for USB2 and USB3, then main hcd's phy
+	* represents the dedicated USB2 phy while shared hcd's phy
+	* represents the USB3 phy which is a different phy than main hcd, we
+	* must power off both phys; otherwise the two hcds shares a same
+	* phy which is for both USB2 and USB3, we only need to power off
+	* the phy once.
+	* Here hcd is the main hcd.
+	*/
+	phy_power_off(hcd->phy);
+	phy_exit(hcd->phy);
+
+	if (of_property_read_bool(dev->of_node, "separated-phys-for-usb2-usb3")) {
+		phy_power_off(xhci->shared_hcd->phy);
+		phy_exit(xhci->shared_hcd->phy);
+	}
+
 	if (!device_may_wakeup(dev) && !IS_ERR(xhci->clk))
 		clk_disable_unprepare(xhci->clk);
 
@@ -451,6 +469,41 @@ static int __maybe_unused xhci_plat_resume(struct device *dev)
 	if (ret)
 		return ret;
 
+	/*
+	* xhci's main hcd is for USB2 while its shared hcd is for USB3;
+	* If phys are separated for USB2 and USB3, then main hcd's phy
+	* represents the dedicated USB2 phy while shared hcd's phy
+	* represents the USB3 phy which is a different phy than main hcd, we
+	* must init and power on both phys; otherwise the two hcds shares
+	* a same phy which is for both USB2 and USB3, we only need to init
+	* and power on the phy once.
+	* Here hcd is the main hcd.
+	*/
+	ret = phy_init(hcd->phy);
+	if (ret)
+		return ret;
+
+	ret = phy_power_on(hcd->phy);
+	if (ret) {
+		phy_exit(hcd->phy);
+		return ret;
+	}
+
+	if (of_property_read_bool(dev->of_node, "separated-phys-for-usb2-usb3")) {
+		ret = phy_init(xhci->shared_hcd->phy);
+		if (ret)
+			return ret;
+
+		ret = phy_power_on(xhci->shared_hcd->phy);
+		if (ret) {
+			phy_exit(xhci->shared_hcd->phy);
+			/* roll back main hcd's phy */
+			phy_power_off(hcd->phy);
+			phy_exit(hcd->phy);
+			return ret;
+		}
+	}
+
 	return xhci_resume(xhci, 0);
 }
 
-- 
1.7.9.5

